core/base/mergeTreeClustering/MergeTreeBase.h

```diff
@@ -0,0 +1,1439 @@
+        ss << i << " _ " << tree->getNode(i)->getOrigin();
+        printMsg(ss.str());
+        printMsg("1 up arc and 1 down arc", debug::Priority::ERROR);
+        myPause();
+      }
+  }
+
+  // For not branch decomposition tree
+  template <class dataType>
+  void putBackMergedNodes(ftm::FTMTree_MT *tree) {
+    bool isJT = tree->isJoinTree<dataType>();
+    std::queue<ftm::idNode> queue;
+    queue.emplace(tree->getRoot());
+    while(!queue.empty()) {
+      ftm::idNode node = queue.front();
+      queue.pop();
+      ftm::idNode nodeOrigin = tree->getNode(node)->getOrigin();
+      if(!tree->isLeaf(node)) {
+        std::vector<ftm::idNode> children = tree->getChildren(node);
+        std::vector<ftm::idNode> lowestNodes;
+        ftm::idNode branchOrigin = nodeOrigin;
+        for(auto child : children) {
+          ftm::idNode lowestNode = tree->getLowestNode<dataType>(child);
+          lowestNodes.push_back(lowestNode);
+          ftm::idNode lowestNodeOrigin = tree->getNode(lowestNode)->getOrigin();
+          if(not tree->isNodeAlone(lowestNodeOrigin)
+             and lowestNodeOrigin != node)
+            branchOrigin = lowestNode;
+        }
+        for(size_t i = 0; i < children.size(); ++i) {
+          ftm::idNode lowestNodeOrigin
+            = tree->getNode(lowestNodes[i])->getOrigin();
+          if(branchOrigin == lowestNodes[i] or lowestNodeOrigin == node)
+            continue;
+          dataType lowestNodeOriginVal
+            = tree->getValue<dataType>(lowestNodeOrigin);
+          ftm::idNode branchOriginT = branchOrigin;
+          ftm::idNode branchRoot = tree->getNode(branchOrigin)->getOrigin();
+          while(branchRoot != branchOriginT) {
+            dataType val
+              = tree->getValue<dataType>(tree->getParentSafe(branchOriginT));
+            if((val > lowestNodeOriginVal and isJT)
+               or (val < lowestNodeOriginVal and not isJT))
+              break;
+            branchOriginT = tree->getParentSafe(branchOriginT);
+          }
+          tree->setParent(lowestNodeOrigin, tree->getParentSafe(branchOriginT));
+          tree->setParent(branchOriginT, lowestNodeOrigin);
+          tree->setParent(children[i], lowestNodeOrigin);
+        }
+      }
+      for(auto child : tree->getChildren(node))
+        queue.emplace(child);
+    }
+  }
+
+  template <class dataType>
+  void postprocessingPipeline(ftm::FTMTree_MT *tree) {
+    // TODO fix dont use min max pair
+    if(not branchDecomposition_ or not useMinMaxPair_)
+      fixMergedRootOrigin<dataType>(tree);
+    if(branchDecomposition_)
+      branchDecompositionToTree<dataType>(tree);
+    else
+      putBackMergedNodes<dataType>(tree);
+  }
+
+  // --------------------------------------------------------------------------------
+  // Output Matching
+  // --------------------------------------------------------------------------------
+  template <class dataType>
+  void computeMatching(
+    ftm::FTMTree_MT *tree1,
```

Comment:
in general, I'd recommend to stick to references by default and only use pointers when you really can't do differently (e.g. when using polymorphism)